### PR TITLE
Implement per-provider tests with lowest-direct dependency resolution

### LINKED ISSUE #35549: Enforce upgrades of min-versions for our dependencies
### Body

It should be entirely possible to check and enforce bumping of min versions for all our dependencies (including providers).

We should be able to run a set of tests (separate test for each provider) where we lower down dependencies tto minimum versions resulting from the resolution and run tests for that provider and see if they pass. Automtating that and making it part of the CI of our should protect us against the case where we add a feature or use an API that is not available in some older version of a dependency we are using.

This would be an ultimate automation of our dependencies where we would care about both sides of the dependency requirements. The current automation cares about upper-binding, but lower-binding is at the discretion of the contributor and keen eye of the reviewer, both are fragile and easy to get wrong - because of the number of dependencies we have, limited review time/focus and sometimes unobvious coupling between versions of dependencies and features/apis.

### Committer

- [X] I acknowledge that I am a maintainer/committer of the Apache Airflow project.

### PR DESCRIPTION
**Pull request by potiuk**
With this change we are running tests of downgrading Airflow
dependencies to "lowest-direct" ones - separately for "core" tests
and for each provider (and run corresponding tests with it). This
should allows us to determine what are the lowest bounds for the
dependencies - for Airflow and for individual providers and continue
doing it while Airflow evolves in the future.

Fixes: https://github.com/apache/airflow/issues/35549

Related: https://github.com/apache/airflow/issues/39100

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

**Comment by JDarDagran** [comment]
Does this PR assume all cross-provider dependencies should be expressed as extras (in terms of python packaging, added in `additional-extras` for each provider)?
Not saying this is a bad approach, just making sure as e.g. OpenLineage provider exists as extra dep only for DBT provider and probably needs manual work to add it as well (or at least have a check in pre-commit if such extra exists, there's already check for affected providers).
`lowest-direct` resolution is a great approach though 👍

**Comment by potiuk** [comment]
> Does this PR assume all cross-provider dependencies should be expressed as extras (in terms of python packaging, added in additional-extras for each provider)?

Unfortunately the way it works in Python, those "extra" limits are not enforceable, and It makes little sense to add them (and verify). There is really no good way to speciffy "if you want to install this optional dependency, it has to follow those limits" - there are multiple scenarios where even with those limits set on extra, user will be able to perform a sequence of upgrades to break it, and at any point in their process even `pip check` will show that everything is fine. So I am not sure we even should attempt to do such a check.

**Comment by JDarDagran** [comment]
> > Does this PR assume all cross-provider dependencies should be expressed as extras (in terms of python packaging, added in additional-extras for each provider)?
> 
> Unfortunately the way it works in Python, those "extra" limits are not enforceable, and It makes little sense to add them (and verify). There is really no good way to speciffy "if you want to install this optional dependency, it has to follow those limits" - there are multiple scenarios where even with those limits set on extra, user will be able to perform a sequence of upgrades to break it, and at any point in their process even `pip check` will show that everything is fine. So I am not sure we even should attempt to do such a check.

Fully agreed. Do you still think that would be worth adding different kind of tests too?
I'll stick to OpenLineage as there have been cases for not keeping backwards-compatibility we did not detect. https://github.com/apache/airflow/pull/39530 also brings such and would love to have something that runs checks against this. My initial idea is to maintain a static list of providers that can be downgraded to the latest released version, similar to how Pydantic and SQLAlchemy have special tests for downgrading. This way, if changes are made to a provider that affect compatibility with OpenLineage, tests will fail if the changes are not backwards-compatible.

**Comment by potiuk** [comment]
Sure - we. An think of next steps and once we get this one merged, you could add a similar follow-up PR with those checks :)

**Comment by potiuk** [comment]
cc: @charliermarsh -> thanks to `uv's` lowest-direct we finally could automate bumping the lower bounding of direct dependencies for Airflow and it's ~90 providers

Also @notatallshaw -> this is probably as close as it can get to the #39100 you opened and likely will make your `apache-airflow[all]` installations quite a bit faster.

File: contributing-docs/testing/unit_tests.rst [hunk 1]
```diff
@@ -1251,6 +1251,131 @@ Rebuilding single provider package can be done using this command:
...
+
+Similarly we can test if the provider tests are working for lowest dependencies of specific provider.
+
+Those tests can be easily run locally with breeze (replace PROVIDER_ID with id of the provider):
+
+.. code-block::bash
+
+    breeze testing tests --force-lowest-dependencies --test-type "Providers[PROVIDER_ID]"
+
+If you find that the tests are failing for some all dependencies, make sure to add minimum version for
```
Comment (aritra24) [inline contributing-docs/testing/unit_tests.rst]: some ~all~ dependencies?

File: contributing-docs/testing/unit_tests.rst [hunk 2]
```diff
@@ -1251,6 +1251,131 @@ Rebuilding single provider package can be done using this command:
...
+ + gcloud-aio-bigquery==6.1.2
+ - gcloud-aio-storage==9.2.0
+
+
+How to fix failing lowest-direct dependency resolution tests
+------------------------------------------------------------
+
+When your tests pass in regular test, but fail in "lowest-direct" dependency resolution tests, you need
+to figure out the lower-bindings missing in  ``hatch_build.py``  (for Airflow core dependencies) or
+in the corresponding provider's ``provider.yaml`` file. This is usually very easy thing that takes a little
```
Comment (aritra24) [inline contributing-docs/testing/unit_tests.rst]: This is usually **a** very easy thing

File: contributing-docs/testing/unit_tests.rst [hunk 3]
```diff
@@ -1251,6 +1251,131 @@ Rebuilding single provider package can be done using this command:
...
+time to figure out especially if you just added new feature from a library that you use, just check in
```
Comment (aritra24) [inline contributing-docs/testing/unit_tests.rst]: little **bit of** time

**Comment by potiuk** [comment]
Also @jscheffl -> in https://github.com/apache/airflow/issues/35549#issuecomment-1804790124 you were concerned that running the tests will take a lot of time. but it's not too much  longer than regular tests even for "full" set of tests that will be run in main.

See: https://github.com/apache/airflow/actions/runs/9328149500/job/25679059016 - it takes about the same time as regular set of complete tests. I did it in a "smart" way - utilising our current parallel test-running infrastructure - each provider tests are run separately and each provider has its own `uv pip install --python /usr/local/bin/python --resolution lowest-direct --editable '.[PROVIDER]' --upgrade` that only downgrads it's dependencies and runs only tests for this provioder, so as result it only takes slightly slower than complete set of tests where we run most Provider tests in a single container.

File: contributing-docs/testing/unit_tests.rst [hunk 4]
```diff
@@ -1251,6 +1251,131 @@ Rebuilding single provider package can be done using this command:
...
+After that, re-run the ``breeze shell --force-lowest-dependencies`` command and see if the tests pass.
+
+.. code-block::bash
+
+   breeze shell --force-lowest-dependencies --test-type "Providers[PROVIDER_ID]"
+
+Sometimes it might get a bit tricky to know what is the minimum version of the library you should be using
+but in this case you can easily find it by looking at the error and list of downgraded packages and
+guessing which one is the one that is causing the problem. You can then look at the release notes of the
+library and find the minimum version but also you can revert to technique knowing as bisecting which allows
```
Comment (aritra24) [inline contributing-docs/testing/unit_tests.rst]: revert to bisecting (or) revert to the technique known as bisecting

**Review by aritra24 — APPROVED** [review]
Few simple nits, can be ignored

**Comment by potiuk** [comment]
Thanks @aritra24. Applied the grammar fixes :)

File: airflow/providers/amazon/provider.yaml [hunk 5]
```diff
@@ -92,17 +92,17 @@ dependencies:
...
   # of candidates to consider. Make sure to configure boto3 version here as well as in all the tools below
   # in the `devel-dependencies` section to be the same minimum version.
-  - boto3>=1.33.0
-  - botocore>=1.33.0
+  - boto3>=1.34.0
+  - botocore>=1.34.0
   - inflection>=0.5.1
   # Allow a wider range of watchtower versions for flexibility among users
-  - watchtower>=2.0.1,<4
+  - watchtower>=3.0.0,<4
```
Comment (potiuk) [inline airflow/providers/amazon/provider.yaml:99]: @o-nikolas @vincbeck @ferruzzi -> I recall that there was a reason you wanted to have watchtower 2, but what I found out during those tests is that **some** tests were failing when it was watchtower 2, so if we go back to watchtower 2, I might need your help on deciding what to do with those and proposing a fix (we could deliberately skip thos tests if watchtower 2 is used or - better implementing a fix to work for both.

**Comment by notatallshaw** [comment]
> Also @notatallshaw -> this is probably as close as it can get to the #39100 you opened and likely will make your `apache-airflow[all]` installations quite a bit faster. UPDATE: Actually you wanted to put limits for airflow -> provider dependencies, so it's a bit differen (those are limits on airlfow and provider's own dependencies) - but let's see where it can get us.

Yeah, it's not quite the same thing, the reason I asked for putting limits on the `airflow -> provider dependencies` is that it exposes resolution algorithms to the possibility of trying to exhaust all versions of a provider dependencies. So while new lower limits on direct dependencies of providers will in general speed up the typical use case, and the extra testing is great, it will not prevent pathalogical backtracking, particularly if the user has additional requirements.

Specifically, if the user is using uv, with Python 3.11, and uv ends up backtracking on the apache-beam provider such they hit these two requirements: `"dill>=0.2.2" "apache-beam<=2.49.0"`, then uv will exhaustively backtrack and try to build an sdist that can not be built, and the resolution will fail.

Fortunately the latest version of `apache-airflow[all]` by itself does not cause that either on main or this branch!

I did a little performance testing with uv, I see  `apache-airflow[all]` resolution speed up by 10% with cache, though without cache the difference was not measurable (uv is too fast on non-pathalogical resolutions so the network cost dominates the amount of time). I will try and get round to testing with pip.

**Review by jscheffl — APPROVED** [review]
Looks good to me.

Spent some time today and iterated around multiple options locally and can confirm all is working... with Python 3.8 I would say LGTM!

...but I also noted that there are more and other problems in other Python versions, I propose after merging this, we need to expand the test to all other Python versions we support as well.. because it seems the dependencies are generating different problems based on used Python version. Just testing against 3.8 does not make all trouble visible.

My tests revealed:

- Python 3.8:
    - apache.impala: Got a timeout in pytest with apache.impala with 60s timeout. Re-run was a success. If such timeouts show in CI we might need to extend timeouts for this test, maybe locally my laptop with 20 cores was a bit overwhelmed.
- Python 3.9:
    - Other Providers[grpc] Providers[hashicorp] Providers[http] Providers[imap] Providers[influxdb]: Failed in 60s timeout - re-test was a success.
    - apache.hdfs: Failing in building dependent wheel package
    - papermill: Failing in building dependent wheel package
    - databricks: Failing in building dependent wheel package (pandas)
- Python 3.10:
    - Other: Failed in 60s timeout - re-test was a success.
    - cloudant: Failed with `ImportError: cannot import name 'Sequence' from 'collections'` in cloudant/_common_util.py:22
    - mongo: Failed with `ImportError: cannot import name 'MutableMapping' from 'collections'` in bson/codec_options.py:19
    - facebook: Failed with `AttributeError: module 'collections' has no attribute 'MutableMapping'` in facebook_business/adobjects/abstractobject.py:29
- Python 3.11:
    - A couple failed in timeouts, but even after re-run the following failures...
    - cloudant, mongo, facebook the same like in Python 3.10
- Python 3.12:
    - Wow, only 1 timeout and runtime is much faster than all other Python versions. Was completed in 22min.
    - papermill: Failed with `ModuleNotFoundError: No module named 'papermill'` .... ah, yeah but this is excluded for Python 3.12 anyway...

**Comment by potiuk** [comment]
> ...but I also noted that there are more and other problems in other Python versions, I propose after merging this, we need to expand the test to all other Python versions we support as well.. because it seems the dependencies are generating different problems based on used Python version. Just testing against 3.8 does not make all trouble visible.

Yep. You are absolutely right - great that you did the check. I have not thought about that one, but you are absolutely right  (and it's easy).

**Comment by potiuk** [comment]
@notatallshaw 

> Fortunately the latest version of apache-airflow[all] by itself does not cause that either on main or this branch!

Yeah. I think the more "airflow" deps are lower-bounded, some of them will have also transitive dependencies that will be limited and we naturally will grow out of the old un-buildable packages.

**Comment by potiuk** [comment]
Actually @jscheffl -> let me run it with all python versions on our self-hosted runners and see :)

File: airflow/providers/amazon/provider.yaml (hunk 5 above)
Comment (potiuk) [reply airflow/providers/amazon/provider.yaml:99]: BTW. Those are the two failing tests when we downgrade watchtower to 2.0.1

```python
____________________________________________________________________ TestCloudwatchTaskHandler.test_write_json_logs[json-serialize] ____________________________________________________________________

self = <Mock name='Queue().put' id='281471674722144'>, args = ({'message': '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": "SomeCustomSerialization(...)"}', 'timestamp': <ANY>},)
kwargs = {}, expected = (({'message': '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": "SomeCustomSerialization(...)"}', 'timestamp': <ANY>},), {})
actual = call({'timestamp': 1717256635862, 'message': '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": null}'})
_error_message = <function NonCallableMock.assert_called_with.<locals>._error_message at 0xffff31e23310>, cause = None

    def assert_called_with(self, /, *args, **kwargs):
        """assert that the last call was made with the specified arguments.

        Raises an AssertionError if the args and keyword args passed in are
        different to the last call to the mock."""
        if self.call_args is None:
            expected = self._format_mock_call_signature(args, kwargs)
            actual = 'not called.'
            error_message = ('expected call not found.\nExpected: %s\nActual: %s'
                    % (expected, actual))
            raise AssertionError(error_message)

        def _error_message():
            msg = self._format_mock_failure_message(args, kwargs)
            return msg
        expected = self._call_matcher((args, kwargs))
        actual = self._call_matcher(self.call_args)
        if expected != actual:
            cause = expected if isinstance(expected, Exception) else None
>           raise AssertionError(_error_message()) from cause
E           AssertionError: expected call not found.
E           Expected: put({'message': '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": "SomeCustomSerialization(...)"}', 'timestamp': <ANY>})
E           Actual: put({'timestamp': 1717256635862, 'message': '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": null}'})

/usr/local/lib/python3.8/unittest/mock.py:913: AssertionError

During handling of the above exception, another exception occurred:

self = <Mock name='Queue().put' id='281471674722144'>, args = ({'message': '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": "SomeCustomSerialization(...)"}', 'timestamp': <ANY>},)
kwargs = {}

    def assert_called_once_with(self, /, *args, **kwargs):
        """assert that the mock was called exactly once and that that call was
        with the specified arguments."""
        if not self.call_count == 1:
            msg = ("Expected '%s' to be called once. Called %s times.%s"
                   % (self._mock_name or 'mock',
                      self.call_count,
                      self._calls_repr()))
            raise AssertionError(msg)
>       return self.assert_called_with(*args, **kwargs)
E       AssertionError: expected call not found.
E       Expected: put({'message': '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": "SomeCustomSerialization(...)"}', 'timestamp': <ANY>})
E       Actual: put({'timestamp': 1717256635862, 'message': '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": null}'})
E
E       pytest introspection follows:
E
E       Args:
E       assert equals failed
E         (                                                                                            (
E           {                                                                                            {
E             'message': '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": null}',                'message': '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": "SomeCustomSerial
E                                                                                                      ization(...)"}',
E             'timestamp': 1717256635862,                                                                  'timestamp': <ANY>,
E           },                                                                                           },
E         )                                                                                            )

/usr/local/lib/python3.8/unittest/mock.py:925: AssertionError

During handling of the above exception, another exception occurred:

self = <tests.providers.amazon.aws.log.test_cloudwatch_task_handler.TestCloudwatchTaskHandler object at 0xffff7c382130>, mock_get_log_events = <MagicMock name='get_log_events' id='281471693451856'>
conf_json_serialize = 'airflow.providers.amazon.aws.log.cloudwatch_task_handler.json_serialize'
expected_serialized_output = '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": "SomeCustomSerialization(...)"}'

    @pytest.mark.parametrize(
        "conf_json_serialize, expected_serialized_output",
        [
            pytest.param(
                "airflow.providers.amazon.aws.log.cloudwatch_task_handler.json_serialize_legacy",
                '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": null}',
                id="json-serialize-legacy",
            ),
            pytest.param(
                "airflow.providers.amazon.aws.log.cloudwatch_task_handler.json_serialize",
                '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": "SomeCustomSerialization(...)"}',
                id="json-serialize",
            ),
            pytest.param(
                None, '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": null}', id="not-set"
            ),
        ],
    )
    @mock.patch.object(AwsLogsHook, "get_log_events")
    def test_write_json_logs(self, mock_get_log_events, conf_json_serialize, expected_serialized_output):
        class ToSerialize:
            def __init__(self):
                pass

            def __repr__(self):
                return "SomeCustomSerialization(...)"

        with conf_vars({("aws", "cloudwatch_task_handler_json_serializer"): conf_json_serialize}):
            handler = self.cloudwatch_task_handler
            handler.set_context(self.ti)
            message = logging.LogRecord(
                name="test_log_record",
                level=logging.DEBUG,
                pathname="fake.path",
                lineno=42,
                args=None,
                exc_info=None,
                msg={
                    "datetime": datetime(2023, 1, 1),
                    "customObject": ToSerialize(),
                },
            )
            with mock.patch("watchtower.threading.Thread"), mock.patch("watchtower.queue.Queue") as mq:
                mock_queue = Mock()
                mq.return_value = mock_queue
                handler.handle(message)
>               mock_queue.put.assert_called_once_with(
                    {"message": expected_serialized_output, "timestamp": ANY}
                )
E               AssertionError: expected call not found.
E               Expected: put({'message': '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": "SomeCustomSerialization(...)"}', 'timestamp': <ANY>})
E               Actual: put({'timestamp': 1717256635862, 'message': '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": null}'})
E
E               pytest introspection follows:
E
E               Args:
E               assert equals failed
E                 (                                                                                            (
E                   {                                                                                            {
E                     'message': '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": null}',                'message': '{"datetime": "2023-01-01T00:00:00+00:00", "customObject": "SomeCustomSerial
E                                                                                                              ization(...)"}',
E                     'timestamp': 1717256635862,                                                                  'timestamp': <ANY>,
E                   },                                                                                           },
E                 )                                                                                            )

tests/providers/amazon/aws/log/test_cloudwatch_task_handler.py:220: AssertionError
```

**Comment by jscheffl** [comment]
> Actually @jscheffl -> let me run it with all python versions on our self-hosted runners and see :). I might want to get it in in this one PR (I am still waiting for the comment from Amazon team about watchtower)

I would have proposed to merge this one first and make another follow-up PR... would have increased the chance to get #40000 for you :-D

**Comment by potiuk** [comment]
> I would have proposed to merge this one first and make another follow-up PR... would have increased the chance to get #40000 for you :-D

I have few other ideas :)

**Comment by potiuk** [comment]
But on a serious note - I really want to hear from Amazon team, because I think the watchtower change might be problematic (can't remember the root cause - so likely we want to fix it before we merge the change).

**Review by eladkal — APPROVED** [review]

File: airflow/providers/amazon/provider.yaml (hunk 5 above)
Comment (eladkal) [reply airflow/providers/amazon/provider.yaml:99]: https://github.com/apache/airflow/pull/35713 is the PR with the reasoning

Comment (eladkal) [reply airflow/providers/amazon/provider.yaml:99]: https://github.com/apache/airflow/pull/35713 is the PR with the reasoning

Comment (potiuk) [reply airflow/providers/amazon/provider.yaml:99]: Yep. I wanted to know if the arguments there are still valid after 6 months. It's pretty natural for those decisions about supporting "past" versions of libraries to get out-of-date and I'd expect maybe 6 months is enough to change the environment to justify bumping watchtower to v3.

**Review by Lee-W — APPROVED** [review]

**Comment by potiuk** [comment]
@jscheffl -> I think I addressed all 3.9 - 3.12 incompatibilities 🤞

File: dev/breeze/src/airflow_breeze/utils/packages.py [hunk 6]
```diff
@@ -443,7 +443,9 @@ def get_install_requirements(provider_id: str, version_suffix: str) -> str:
         dependencies = get_provider_requirements(provider_id)
     else:
         dependencies = PROVIDER_DEPENDENCIES.get(provider_id)["deps"]
-    install_requires = [apply_version_suffix(clause, version_suffix) for clause in dependencies]
+    install_requires = [
+        apply_version_suffix(clause, version_suffix).replace('"', '\\"') for clause in dependencies
```
Comment (jscheffl) [inline dev/breeze/src/airflow_breeze/utils/packages.py:446]: Does the python version in the dependency actually _need_ to be quoted? As the version is also w/o quotes...

Comment (potiuk) [reply dev/breeze/src/airflow_breeze/utils/packages.py:446]: Yes

Comment (potiuk) [reply dev/breeze/src/airflow_breeze/utils/packages.py:446]: https://peps.python.org/pep-0508/#environment-markers

File: airflow/providers/amazon/provider.yaml (hunk 5 above)
Comment (vincbeck) [reply airflow/providers/amazon/provider.yaml:99]: > @o-nikolas @vincbeck @ferruzzi -> I recall that there was a reason you wanted to have watchtower 2, but what I found out during those tests is that **some** tests were failing when it was watchtower 2, so if we go back to watchtower 2, I might need your help on deciding what to do with those and proposing a fix (we could deliberately skip thos tests if watchtower 2 is used or - better implementing a fix to work for both.

Sounds good, I cannot recall the reason why we needed watchtower 2. Let's try with 3 and see it from there

File: .github/workflows/run-unit-tests.yml [hunk 7]
```diff
@@ -140,6 +151,7 @@ jobs:
       PYDANTIC: "${{ inputs.pydantic }}"
       PYTHON_MAJOR_MINOR_VERSION: "${{ matrix.python-version }}"
       UPGRADE_BOTO: "${{ inputs.upgrade-boto }}"
+      AIRFLOW_MONITOR_DELAY_TIME_IN_SECONDS: "${{inputs.monitor-delay-time-in-seconds}}"
```
Comment (ferruzzi) [inline .github/workflows/run-unit-tests.yml:154]: Jinja is not my forte, so maybe it doesn't matter or maybe there is a reason it is different here, but the other lines around this have spaces between the curly braces and the text inside them.

```suggestion
      AIRFLOW_MONITOR_DELAY_TIME_IN_SECONDS: "${{ inputs.monitor-delay-time-in-seconds }}"
```

Comment (potiuk) [reply .github/workflows/run-unit-tests.yml:154]: It's not JINJA - it's GitHub Actions expression language :). It does not matter really, if spaces are there - would be good for consistency, but I think we already use it differently in different places .. as long as we won't be able to have some linting to add it, it does not bother me too much :)

File: airflow/providers/amazon/provider.yaml (hunk 5 above)
Comment (potiuk) [reply airflow/providers/amazon/provider.yaml:99]: Cool.

**Comment by potiuk** [comment]
And merged :)

**Comment by potiuk** [comment]
BTW. When it comes to `--lowest` strategy, we might want to wait to try it when https://github.com/astral-sh/uv/pull/4006 gets merged - then we will see clearer which packages might cause problems there - but I am afraid it will get things "too much" to the floor, and we will be fixing someone else's problems this way.

**Comment by simonprydden** [comment]
> But on a serious note - I really want to hear from Amazon team, because I think the watchtower change might be problematic (can't remember the root cause - so likely we want to fix it before we merge the change).

@potiuk  did you hear back from the MWAA team? 

The issue is that MWAA pins watchtower==2.0.1 , as per [docs](https://docs.aws.amazon.com/mwaa/latest/userguide/connections-packages.htmll)

> Note
> For Apache Airflow v2 and above, Amazon MWAA installs [Watchtower version 2.0.1](https://pypi.org/project/watchtower/2.0.1/) after performing pip3 install -r requirements.txt, to ensure compatibility with CloudWatch logging is not overridden by other Python library installations.

Trying to find this PR in the amazon providers changelog, to double check which is the max version of the amazon providers that will work with MWAA. I think it's 8.24.0

**Comment by potiuk** [comment]
> @potiuk did you hear back from the MWAA team?

Yes. See  https://github.com/apache/airflow/pull/39946#discussion_r1623240120 and maybe @vincbeck or @ferruzzi  or @o-nikolas can comment on that :)

**Comment by o-nikolas** [comment]
> > @potiuk did you hear back from the MWAA team?
> 
> Yes. See [#39946 (comment)](https://github.com/apache/airflow/pull/39946#discussion_r1623240120) and maybe @vincbeck or @ferruzzi or @o-nikolas can comment on that :)

I've asked around on the MWAA team and I'll report what I hear back!

**Pull request merged**